diff --git a/aio.el b/aio.el
index 203f441..e6ac57a 100644
--- a/aio.el
+++ b/aio.el
@@ -24,2 +24,2 @@
 (require 'cl-lib)
 (require 'generator)

+(eval-and-compile
+  ;; Tell the native compiler about the types defined with `record`.
+  (when (and (fboundp 'native-comp-available-p)
+             (native-comp-available-p))
+    (cl-defstruct aio-promise)
+    (cl-defstruct aio-select)
+    (cl-defstruct aio-sem)
+    (put 'aio-promise 'comp-native-type-proper t)
+    (put 'aio-select 'comp-native-type-proper t)
+    (put 'aio-sem 'comp-native-type-proper t)))
+
 ;; Register new error types
 (define-error 'aio-cancel "Promise was canceled")
 (define-error 'aio-timeout "Timeout was reached")
